Update Seq.md with small tweaks, one expanded explanation (#3015)
ocaml/ocaml.org: data/tutorials/language/3ds_05_seq.md

## Introduction

Sequences are very much like lists. However, from a pragmatic perspective, one

> should imagine they are infinite. That's the key intuition to understanding

Reviewer: ah, the thing is sequences can be finite, so they are not always infinite. Worth to make sure that this is captured, and it would be fine to rewrite the entire paragraph to be more clear.

Author: Good catch, thank you :)

> should imagine they are infinite. That's the key intuition to understanding

Reviewer 2: Because `Seq.Node` has **two** constructors: `Nil` and `Cons`, a sequence _may_ be finite. This defines a type of strictly infinite sequences:
```ocaml
module Stream : sig
  type 'a t = unit -> 'a node
  and 'a node = Cons of 'a * 'a t 
end
```
We have three cases:
* `list`: strictly finite
* `Stream.t`: strictly infinite
* `Seq.t`: finite or infinite

This tutorial would be easier to write if `Stream` was part of the standard library, unfortunately, it isn't. Arguably there are few use cases for `Stream` (except teaching).

Author: Thats an interesting way to look at it. I'll put Stream on the list of potential lessons

Reviewer 2: We have this: https://ocaml.org/exercises#100

[_thunk_](https://en.wikipedia.org/wiki/Thunk). Using this terminology, `Seq.t`
values are thunks. With the analogy used earlier, `a` is frozen in its thunk.

## Constructing Sequences

With this understanding, we can manually construct a sequence like so:

``` ocaml
let my_seq  =
  fun () ->
  Seq.Cons (1, fun () -> Seq.Cons (2, fun () -> Seq.Cons (3, fun () -> Seq.Nil)))
```

**Note:** The second component of each `Seq.Con`'s tuple is a function. This has
the effect of providing a means of acquiring a value rather than providing a
value directly.

We can also construct sequences using functions. Here is how to build seemingly

> infinite sequences of integers:

Reviewer: ````suggestion
## Constructing Sequences

With this understanding, we can manually construct a sequence like so:

``` ocaml
let my_seq  =
  fun () ->
  Seq.Cons (1, fun () -> Seq.Cons (2, fun () -> Seq.Cons (3, fun () -> Seq.Nil)))
```

**Note:** The second component of each `Seq.Con`'s tuple is a function. This has
the effect of providing a means of acquiring a value rather than providing a
value directly.

We can also construct sequences using functions. Here is how to build an
infinite sequence of integers:
````

Love the addition of showing how to define a finite sequence because it gives more intuition how sequence works.

The function `ints n` looks as if building the infinite sequence `(n; n + 1; n +
2; n + 3;...)`. In reality, since machine integers have bounds, the sequence
isn't indefinitely increasing. For technical reasons, when `max_int` is reached, it will circle
down to `min_int`.

The OCaml Standard Library contains a module for sequences called

> [`Seq`](/releases/5.0/api/Seq.html). It contains `Seq.int`, which we implemented above.

Reviewer 2: ```suggestion
[`Seq`](/releases/api/Seq.html). It contains `Seq.int`, which we implemented above.
```

 509; 521; 523; 541]
```

The function `trial_div` is recursive in OCaml and can be understood if we break
it down into its constituent parts. It is defined using the `rec` keyword, allowing the
function to call itself. For each loop in the recursive call, it pattern-matches
on either `Seq.Cons (m, seq)` or the end of the sequence, `Seq.Nil`.

> If it matches on the first branch (`Seq.Cons (m, seq)`), we filter the

Reviewer 2: ```suggestion
If it matches on the first branch `Seq.Cons (m, seq)`, we filter the
```

remaining sequence of all integers that are divisible by `m` before recursively
calling `trial_div` on the filtered sequence. This branch is matched on until we
reach the end of the sequence for every recursive call.

So far, we recursively traveled down our sequence until we reached the 100th
prime number. Next, we retrace our steps up the recursive trail, wherein we
construct our result by calling `Seq.Cons` on `m` and the previously constructed
filtered sequence beginning with `Seq.Nil`.

> **Side Note**: It may be interesting to learn that `trial_div`, while it can
> colloquially be called a recursive, is an example of a kind of recursion called
> [corecursive](https://en.wikipedia.org/wiki/Corecursion). Corecursion differs
> from recursion in that it constructs values incrementally rather than computing
> a final result all at once. Unlike traditional recursion, which works towards a
> base case, corecursive functions can produce an infinite stream of values,
> generating only as much as needed. The `trial_div` function is corecursive
> because it does not compute the full sequence of primes immediately. Instead, it
> produces prime numbers lazily, filtering and deferring further computation until
> more elements are requested. This allows the sequence to be processed
> incrementally rather than requiring a complete traversal upfront.

Reviewer 2: ```suggestion
**Side Note**: It may be interesting to learn that `trial_div`, while it can
colloquially be called a recursive, is an example of a kind of recursion called
[corecursion](https://en.wikipedia.org/wiki/Corecursion). Corecursion differs
from recursion in that it constructs results incrementally rather than consuming
it's input incrementally. Unlike traditional recursion, which works towards a
base case, corecursive functions must indefinitely produce values as a stream. The `trial_div` function is corecursive
because it does not immediately compute the complete sequence of primes. Instead, it
produces prime numbers on-demand, filtering and deferring further computation until
more elements are requested. This allows the sequence to be processed
incrementally rather than requiring a complete traversal upfront.
```

The function `Seq.uncons` returns the head and tail of a sequence if it is not
empty. Otherwise, it returns `None`.

### Reading a File with `Seq.Unfold`

For the next example, we will demonstrate the versatility of `Seq.unfold` by
using it to read a file.

> Before doing so, lets define a function that processes a file's contents from a
> provided channel:

Reviewer 2: ```suggestion
Before doing so, let's define a function that reads a file's line from a
provided channel, with the type signature needed by `Seq.unfold`.
```

```ocaml
# let input_line_opt chan =
    try Some (In_Channel.input_line chan, chan)
    with End_of_file -> None;;
val input_line_opt : in_channel -> (string * in_channel) option = <fun>
```

**Note**: To make the code in the next section work, create a file named "README.md" and add dummy content. We use a file generated by the following command:

``` shell
cat > README.md <<EOF
This is the first line.
This is the second line.
EOF
```

Finally, lets define a variable that holds the file's contents and is

> constructed using `Seq.unfold`:

Reviewer 2: ```suggestion
Finally, let's read the file's contents using `Seq.unfold`. Mind that `cin` is a local definition.
```

Mind that _variable_ is in our list of banned words, see [`DOCUMENTATION_WRITING.md`](../../blob/main/DOCUMENTATION_WRITING.md)

Stack overflow during evaluation (looping recursion?).
```

This definition is behaving as expected (spot the differences, there are four): <!-- How do you count four?

> 1. Seq.Cons vs Seq.cons

Reviewer 2: ```suggestion
1. `Seq.Cons` vs `Seq.cons`
```

with all the arguments it expects. In the latter definition, the application is
partial because the `()` argument is missing. Since evaluation is
[eager](https://en.wikipedia.org/wiki/Evaluation_strategy#Eager_evaluation) in
OCaml, in the former case evaluation of the recursive call is triggered and

> a non-terminating looping occurs. By contrast in the latter case, the partially

Reviewer: ```suggestion
OCaml, in the former case evaluation of the recursive call is triggered again and again, without ever terminating (this is what "looping recursion" in the error message refers to). In the latter case, the partially
```
